File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {

        List<List<Integer>> graph = new ArrayList<>();
        int[] indegree = new int[numCourses];

        for(int i=0; i<numCourses; i++){
            graph.add(new ArrayList<Integer>());
        }

        for(int[] preq : prerequisites){
            graph.get(preq[1]).add(preq[0]);
            indegree[preq[0]]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for(int i=0; i<numCourses; i++){
            if(indegree[i] == 0){
                queue.offer(i);
            }
        }

        int availableCount = 0;

        while(!queue.isEmpty()){
            int currNum = queue.poll();
            availableCount++;

            for(int neighbor : graph.get(currNum)){
                indegree[neighbor]--;

                if(indegree[neighbor] == 0){
                    queue.offer(neighbor);
                }
            }

        }

        return availableCount == numCourses;
    }
}

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        graph = [ [] for _ in range(numCourses)]
        visited = [ 0 for _ in range(numCourses)]

        for end, start in prerequisites:
            graph[start].append(end)

        
        for num in range(numCourses):
            if not self.dfs(graph, visited, num):
                return False
        
        return True

    
    def dfs(self, graph, visited, num):

        if visited[num] == -1:
            return False
        
        if visited[num] == 1:
            return True

        visited[num] = -1

        for neighbor in graph[num]:
            if not self.dfs(graph, visited, neighbor):
                return False

        visited[num] = 1        

        return True
